src/process.rs

```diff
@@ -637,15 +636,333 @@ mod unix {
+        /// received via the environment.
+        ///
+        /// If found, removes the file descriptor from the collection, sets
+        /// the FD_CLOEXEC flag on the file descriptor and returns it as the
+        /// Rust type Some(UdpSocket).
+        fn pop<T: FromRawFd>(&mut self, ty: SocketType) -> Option<T> {
+            self.fds
+                .iter()
+                .position(|v| v.socket_type == ty)
+                .and_then(|idx| self.fds.remove(idx).finalize())
+        }
+    }
+
+    /// An internal data structure for storing information about sockets
+    /// received via environment variables.
+    #[derive(Debug)]
+    struct SocketInfo {
+        /// The type of socket.
+        pub socket_type: SocketType,
+
+        /// The address of the socket.
+        pub socket_addr: SocketAddr,
+
+        /// The underlying socket file descriptor.
+        pub raw_fd: RawFd,
+    }
+
+    impl SocketInfo {
+        /// Creates a new [`SocketInfo`] instance.
+        fn new(socket_type: SocketType, socket_addr: SocketAddr, raw_fd: RawFd) -> Self {
+            Self {
+                socket_type,
+                socket_addr,
+                raw_fd,
+            }
+        }
+
+        /// Per [`sd_listen_fds()`] set the FD_CLOEXEC flag on the returned
+        /// socket.
+        ///
+        /// Consumes self.
+        ///
+        /// Returns Some(T) if the FD_CLOEXEC flag could be set, None
+        /// otherwise.
+        fn finalize<T: FromRawFd>(self) -> Option<T> {
+            log::trace!("Setting FD_CLOEXEC on SystemD LISTEN_FDS received socket file descriptor {}", self.raw_fd);
+            match fcntl(self.raw_fd, FcntlArg::F_SETFD(FdFlag::FD_CLOEXEC)) {
+                Ok(_) => unsafe {
+                    return Some(FromRawFd::from_raw_fd(self.raw_fd));
+                }
+                Err(err) => {
+                    log::warn!("Setting FD_CLOEXEC on SystemD LISTEN_FDS received socket file descriptor {} failed: {err}", self.raw_fd);
+                }
+            }
+            None
+        }
+
+        /// Wrap a socket file descriptor into a SocketInfo instance,
+        /// if we support it.
+        ///
+        /// Supported file descriptors:
+        ///   - Represent UDP or TCP sockets.
+        ///   - Have an address.
+        ///
+        /// Returns Some(SocketInfo) on success, None otherwise.
+        fn from_fd(fd: i32) -> nix::Result<SocketInfo> {
+            // [`getsockname()`]` will fail if the given argument is not "a
+            // valid file descriptor" or does not "refer to a socket", so we
+            // don't need to call fstat() to check that the FD is a socket, we
+            // can let getsockname() take care of that for us.
+            //
+            // [`getsockname()`]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/getsockname.html#tag_16_219_05
+            let sock_addr = getsockname::<SockaddrStorage>(fd)?;
```

Comment:
Of course this only works as long as TCP/UDP sockets is all you want. There may be a use case for Unix sockets and, especially, pipes, though.

Reply:
Yes, I didn't need that at present so didn't add it.

Reply:
Another thing to worry about later, I suppose.